### DISCUSSION TITLE
g.V().count() returning very low count ?

### DISCUSSION THREAD

**Discussion by hexa00** [JanusGraph/janusgraph#3378] [Q&A] [answered]
Hi,
   Is it normal for Janus using a Cassandra backend to return very low vertex counts when using g.V().count() for example:
```
[Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
JanusGraphStep([])                                                     9           9          13.656    96.04
  constructGraphCentricQuery                                                                   0.020
  GraphCentricQuery                                                                           13.461
    \_condition=()
    \_isFitted=false
    \_query=[]
    \_orders=[]
    \_isOrdered=true
    scan                                                                                      13.362
      \_condition=VERTEX
      \_query=[]
      \_fullscan=true
CountGlobalStep                                                        1           1           0.562     3.96
                                            >TOTAL                     -           -          14.218        -]
````

However there's 100k + vertices in that graph ...

**Comment by vtslab** [comment]
No, that is not normal. Please check other scenarios. E.g. if you use one JanusGraph instance to add 100k vertices, a separate JanusGraph instance cannot see them until they are committed to the backends.
If you would happen to encounter a bug, we would need a scenario to reproduce it.

**Reply by hexa00** [reply]
OK yeah the vertices are in the backend for sure, they can be returned by hitting the mixed index and they have been ingested for a while.

Maybe this has something to do with our use of keyspaces, I'm thinking we'll trace the CQL queries and see what's going on, hopefully that helps.

**Comment by hexa00** [comment, marked as answer]
Requests for g.V().count()
```
437208 [gremlin-server-exec-6] WARN  org.janusgraph.graphdb.transaction.StandardJanusGraphTx  - Query requires iterating over all vertices [()]. For better performance, use indexes
437215 [JanusGraph Session-io-2] INFO  com.datastax.oss.driver.internal.core.tracker.RequestLogger  - [JanusGraph Session|999836508][Node(endPoint=3.12.23.184/3.12.23.184:9142, hostId=ac9901f0-287c-3994-9da3-3eeec9af4705, hashCode=28184da4)] Success (6 ms) [2 values] SELECT key,column1,value FROM graph.edgestore WHERE column1>=:slicestart AND column1<=:sliceend ALLOW FILTERING [slicestart=0x02, sliceend=0x03]
437224 [JanusGraph Session-io-3] INFO  com.datastax.oss.driver.internal.core.tracker.RequestLogger  - [JanusGraph Session|2124075595][Node(endPoint=3.12.23.188/3.12.23.188:9142, hostId=bbc8d4cc-caa3-3b7e-ba41-c5c01169de46, hashCode=5a20bf89)] Success (8 ms) [2 values] SELECT key,column1,value FROM graph.edgestore WHERE column1>=:slicestart AND column1<=:sliceend ALLOW FILTERING [slicestart=0x02, sliceend=0x03]
437230 [JanusGraph Session-io-0] INFO  com.datastax.oss.driver.internal.core.tracker.RequestLogger  - [JanusGraph Session|1321987316][Node(endPoint=3.12.23.137/3.12.23.137:9142, hostId=0497d827-22ae-3776-9b9d-14a60347756f, hashCode=197a02ed)] Success (5 ms) [2 values] SELECT key,column1,value FROM graph.edgestore WHERE column1>=:slicestart AND column1<=:sliceend ALLOW FILTERING [slicestart=0x02, sliceend=0x03]
437234 [JanusGraph Session-io-1] INFO  com.datastax.oss.driver.internal.core.tracker.RequestLogger  - [JanusGraph Session|1328466181][Node(endPoint=REDACTED/3.12.23.142:9142, hostId=05deae2d-6405-494d-a965-c0e5836bcb3c, hashCode=30518d40)] Success (4 ms) [2 values] SELECT key,column1,value FROM graph.edgestore WHERE column1>=:slicestart AND column1<=:sliceend ALLOW FILTERING [slicestart=0x02, sliceend=0x03]
437243 [JanusGraph Session-io-2] INFO  com.datastax.oss.driver.internal.core.tracker.RequestLogger  - [JanusGraph Session|915777815][Node(endPoint=3.12.23.171/3.12.23.171:9142, hostId=b5bff396-4e0c-3939-a5c6-64d66d198d69, hashCode=22a8ae77)] Success (8 ms) [2 values] SELECT key,column1,value FROM graph.edgestore WHERE column1>=:slicestart AND column1<=:sliceend ALLOW FILTERING [slicestart=0x02, sliceend=0x03]

```

**Reply by hexa00** [reply]
Maybe a problem with the slicing?

**Reply by li-boxuan** [reply]
Slicing is right. How many results do you see if you do `g.V().toList()`? Also, you said 

> they can be returned by hitting the mixed index and they have been ingested for a while.

Can you show your query and results? It might be the case that the data is already deleted but mixed index backend still has the stale data.

**Reply by hexa00** [reply]
```
g.V().count()
```
```
[11]
```
There's nothing I think that would delete the data ? We're not deleting for sure on purpose.

Also if the mixed index was stale it would not return the properties included in the vertex ? In my case I get all the data for those vertices...

**Reply by hexa00** [reply]
```
g.V().toList()
```
```
[v[3940568], v[68534360], v[23908352], v[7467192], v[23494824], v[12382232], v[23195784], v[52158536], v[5132312], v[21573768], v[63639584]]
```

**Reply by hexa00** [reply]
```
g.V().has('somekey_onmixed', 'value').limit(100).toList()
```
```
[v[61576], v[127120], v[41369600], v[41058496], v[249912], v[213160], v[82071720], v[213216], v[258216], v[307256], v[315448], v[262176], v[606376], v[53432], v[41184], v[41066688], v[81944576], v[94216], v[282664], v[41500672], v[143448], v[82059264], v[45184], v[73896], v[147528],... for 100 vertices
```

**Reply by hexa00** [reply]
I can also see all the data for those

**Reply by li-boxuan** [reply]
Try 

`g.V().has('somekey_onmixed', 'value').limit(100).valueMap().toList()`

**Reply by hexa00** [reply]
I can see all the data:
```
[{var0=[XXX], var1=[vvvv], var2=[xxx],...}, {var0=[xxx] ... etc..
```
Sorry I can't paste the actual data, but I see all the values mapped properly and their data.

**Reply by li-boxuan** [reply]
Can you try `SELECT COUNT(*) FROM graph.edgestore` in AWS Keyspace CQL editor?

**Reply by hexa00** [reply]
I get :
```Invalid query
countRows is not yet supported.
```
Doing that but if I try just `SELECT *`  I get 100 results (likely there's a 100 limit)

**Reply by hexa00** [reply]
Tables shows about 600MB used too...

**Reply by hexa00** [reply]
I've tried to restart by DB without TTL enabled at all, so far I get the expected number of vertices... humm
I'm going to let it ingest more data a bit, then switch the TTL on and see if it goes down

**Reply by li-boxuan** [reply]
Cool. Please share what you find out. It might be an AWS keyspace issue as well.

In case you still see rows missing, try the following query:

`SELECT key,column1,value FROM graph.edgestore WHERE column1>=0x02 AND column1<=0x03 ALLOW FILTERING`

FYI, this [blog](https://medium.com/@li-boxuan/janusgraph-deep-dive-data-layout-in-janusgraph-cd33c045a495) explains the data stored on the backend. 

Basically, a row with `column1 = 0x02` represents the existence of a vertex.

**Reply by hexa00** [reply]
I did that on the keyspace I am observing the problem on and I get hundreds of values...
Seems weird because that seems to be the same query that the tracing was showing except the number of results really doesn't match

**Reply by hexa00** [reply]
Trying out `SELECT key,column1,value FROM graph.edgestore WHERE column1>=0x02 AND column1<=0x03 ALLOW FILTERING` in a CQL editor with the keyspace with TTL disabled I get 7k + Rows, 

While the count says 1542 , actually just doing that query over and over it goes up and down 2000, 950, 1400, seems a bit random, while we do no deletes on our end...

The mixed index shows about 4200 vertices meanwhile :(

**Reply by li-boxuan** [reply]
What's the JanusGraph version you are using?

> SELECT key,column1,value FROM graph.edgestore WHERE column1>=0x02 AND column1<=0x03 ALLOW FILTERING

Can you paste a few random records (ideally, a few from the beginning, a few from the middle, and a few from the end) here? That shouldn't contain any sensitive information because it only records the vertex id and meta-data about its existence. Of course, if you are not comfortable sharing it publicly, but you can share it with me privately, that would also work. You can find my email on my GitHub profile.

**Reply by hexa00** [reply]
I'm using 0.6.2

Here's some data:
```
"column1","value","key"
"Ag==","AAECNlCb","2AAAAAAB4QA="
"Ag==","AAEEClyO","cAAAAAACo4A="
"Ag==","AAEUMnif","'+AAAAAAT+IA="
"Ag==","AAEWSRyE","IAAAAAAVVYA="
"Ag==","AAEDI1yb","2AAAAAACnQA="
"Ag==","AAEFXnCJ","SAAAAAADgQA="
"Ag==","AAEWTFiE","IAAAAAAVXIA="
"Ag==","AAEBZgCB","CAAAAAAA8YA="
"Ag==","AAEBanyP","eAAAAAABZ4A="
"Ag==","AAEVCjyB","CAAAAAAUR4A="
"Ag==","AAEWNVyC","EAAAAAAVrQA="
....
"Ag==","AAECNlCb","2AAAAAAB4QA="
"Ag==","AAEEClyO","cAAAAAACo4A="
"Ag==","AAEUMnif","'+AAAAAAT+IA="
"Ag==","AAEWSRyE","IAAAAAAVVYA="
"Ag==","AAEDI1yb","2AAAAAACnQA="
"Ag==","AAEFXnCJ","SAAAAAADgQA="
"Ag==","AAEWTFiE","IAAAAAAVXIA="
"Ag==","AAEBZgCB","CAAAAAAA8YA="
"Ag==","AAEBanyP","eAAAAAABZ4A="
"Ag==","AAEVCjyB","CAAAAAAUR4A="
"Ag==","AAEWNVyC","EAAAAAAVrQA="
"Ag==","AAFYeII=","EAAAAAAAeQA="
"Ag==","AAEXOWCC","EAAAAAAWdwA="
"Ag==","AAECXUSS","kAAAAAAB+wA="
"Ag==","AAECYmyZ","yAAAAAAB5wA="
"Ag==","AAEVbzyc","4AAAAAAU7QA="
"Ag==","AAE+SIo=","UAAAAAAALoA="
"Ag==","AAEKOniO","cAAAAAAHoYA="
"Ag==","AAECF1SK","UAAAAAABUYA="
"Ag==","AAEZEBSF","KAAAAAAWzAA="
"Ag==","AAEDBnCB","CAAAAAABygA="
"Ag==","AAEBbRiI","QAAAAAABYgA="
"Ag==","AAEENJk=","yAAAAAAABoA="
"Ag==","AAECDSyJ","SAAAAAAA2oA="
...
"Ag==","AAETcUyI","QAAAAAATzwA="
"Ag==","AAEfKIo=","UAAAAAAAKAA="
"Ag==","AAEQKIg=","QAAAAAAAEgA="
"Ag==","AAEDdwSE","IAAAAAAC1QA="
"Ag==","AAEEWyiH","OAAAAAADAwA="
"Ag==","AAETWXSI","QAAAAAATogA="
"Ag==","AAEEDCSO","cAAAAAACpoA="
"Ag==","AAEVeHSC","EAAAAAAVQ4A="
"Ag==","AAEUVWic","4AAAAAAUJ4A="
"Ag==","AAECBTyS","kAAAAAABWYA="
"Ag==","AAFqYIc=","OAAAAAAAPgA="
"Ag==","AAE8EDCA","AAAAAAABEoA="
"Ag==","AAEXEUSC","EAAAAAAWNAA="
"Ag==","AAEbYIU=","KAAAAAAAIAA="
"Ag==","AAEDTmCN","aAAAAAACqwA="
"Ag==","AAEJaAyO","cAAAAAAHPAA="
"Ag==","AAFPKzyA","AAAAAAAUFoA="
"Ag==","AAECDhiN","aAAAAAABZQA="
"Ag==","AAEEC3iQ","gAAAAAACuYA="
"Ag==","AAEHOmSE","IAAAAAAFnAA="
"Ag==","AAEWLDiB","CAAAAAAVMYA="
"Ag==","AAESGI0=","aAAAAAAAEoA="
"Ag==","AAEBKnyE","IAAAAAAAugA="
"Ag==","AAEBPgSN","aAAAAAAA3AA="
"Ag==","AAECYDCO","cAAAAAABkIA="
```

I appreciate the time you're taking helping with this!

**Reply by hexa00** [reply]
Note I could not get to the end...
It's from results at about 1-100, then 300-400, and 700-800

**Reply by hexa00** [reply]
Also interestingly now g.V().count() is 59... seems like it goes down with time...
While the mixed index count is 50 000

**Reply by li-boxuan** [reply]
What partitioner are you using? You can find it on `Amazon Keyspaces > Configuration`.

What's the `storage.cql.partitioner-name` you have in your config file? AWS keyspace used to only support their own `DefaultPartitioner` but nowadays they change the default setting to `Murmur3Partitioner`.

**Reply by hexa00** [reply]
It's `storage.cql.partitioner-name=DefaultPartitioner` indeed I can try changing that

**Reply by hexa00** [reply]
Doesn't look like it had any effect...

**Reply by li-boxuan** [reply]
What did you change? Can you try changing it back to DefaultPartitioner on AWS keyspace UI and wait for 10 minutes?

> Note that the configuration change takes approximately 10 minutes to complete. 

I don't think this is the cause but worth trying.

**Reply by hexa00** [reply]
oh oops I changed the JanusConfig but not AWS

**Reply by hexa00** [reply]
Yeah I can't change that atm since it's an account level settings and there's others it would affect ;(

**Reply by hexa00** [reply]
I really don't get how the same query basically from Janus
```
437215 [JanusGraph Session-io-2] INFO  com.datastax.oss.driver.internal.core.tracker.RequestLogger  - [JanusGraph Session|999836508][Node(endPoint=3.12.23.184/3.12.23.184:9142, hostId=ac9901f0-287c-3994-9da3-3eeec9af4705, hashCode=28184da4)] Success (6 ms) [2 values] SELECT key,column1,value FROM graph.edgestore WHERE column1>=:slicestart AND column1<=:sliceend ALLOW FILTERING [slicestart=0x02, sliceend=0x03]
```
And manually doing the same would yield different results...

**Reply by li-boxuan** [reply]
Okay never mind, let me check a few things about your data and come back to you.

Another thing worth trying is to use a Cassandra client (e.g. https://github.com/datastax/java-driver) and execute the same query I mentioned above.

**Reply by hexa00** [reply]
Indeed, might try to run janus in debug at that point.

**Reply by hexa00** [reply]
Note I manged to connect using cqlsh
It gives a better format for the data
```
qlsh> SELECT key,column1,value FROM entity_graph_janusgraph.edgestore WHERE column1>=0x02 AND column1<=0x03 ALLOW FILTERING;

 key                | column1 | value
--------------------+---------+----------------
 0xd80000000001e100 |    0x02 | 0x00010236509b
 0x700000000002a380 |    0x02 | 0x0001040a5c8e
 0xf80000000013f880 |    0x02 | 0x00011432789f
 0x2000000000155580 |    0x02 | 0x000116491c84
 0xd800000000029d00 |    0x02 | 0x000103235c9b
....
 key                | column1 | value
--------------------+---------+----------------
 0x0800000000013d00 |    0x02 | 0x0001021e2c81
 0x7000000000017000 |    0x02 | 0x0001024e148e
 0x800000000003ea00 |    0x02 | 0x000105703890
 0xc000000000014c00 |    0x02 | 0x0001017d5098
 0xc000000000004d80 |    0x02 |   0x00014f7498
 0xc800000000017f80 |    0x02 | 0x000102285899
 0x6000000000005680 |    0x02 |   0x0001375c8c

--MORE---
 key                | column1 | value
--------------------+---------+----------------
 0xd000000000000a80 |    0x02 |   0x000109489a
 0x0800000000145880 |    0x02 | 0x000115152c81
 0xe000000000026c00 |    0x02 | 0x0001034d689c
 0x7000000000071200 |    0x02 | 0x0001094f608e
 0x280000000016c300 |    0x02 | 0x0001190a7485
 0x280000000015a500 |    0x02 | 0x0001172c5885
 0xc00000000001dc00 |    0x02 | 0x0001024c3898

---MORE---
---MORE---
---MORE---
---MORE---
---MORE---

```
Note the empty  ---MORE-- there I wonder if some batches come up empty... 
Could it be Janus stops ahead of time as it receives an empty batch ?....

**Reply by li-boxuan** [reply]
> Could it be Janus stops ahead of time as it receives an empty batch

That sounds like a reasonable doubt. To validate that, you would need to use the DataStax java driver since that is the one JanusGraph uses.

Btw I just created a JanusGraph instance on AWS keyspace, and inserted 10k nodes, but couldn't reproduce your problem. Your data looks fine to me too.

**Reply by li-boxuan** [reply]
You could also use an IDE to launch JanusGraph and set debugger points to see what happens. Below is a (subset) of call stacks (from top to bottom) which you can inspect:

```
CQLKeyColumnValueStore::next
CQLKeyColumnValueStore::getKeys(final SliceQuery query, final StoreTransaction txh)
StandardJanusGraph::getVertexIDs
VertexIterable.java
StandardJanusGraphTx::getInternalVertices
```

I suspect Amazon Keyspace's behavior diverges from standard Cassandra and this causes the java driver to misinterpret the results.

**Reply by li-boxuan** [reply]
If you use https://github.com/datastax/java-driver and see different results from running CQL on AWS Keyspace UI, it should be a bug of AWS Keyspace and you should report it to AWS.

**Reply by hexa00** [reply]
Yeah since you can't reproduce, maybe there's something wrong with my setup somehow.  adding my conf in case you see something weird ?
```
# NOTE: THIS FILE IS GENERATED VIA "update.sh"
# DO NOT EDIT IT DIRECTLY; CHANGES WILL BE OVERWRITTEN.
#
# Copyright 2021 JanusGraph Authors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

gremlin.graph = org.janusgraph.core.JanusGraphFactory

# JanusGraph configuration sample: Cassandra & Elasticsearch over sockets
#
# This file connects to Cassandra and Elasticsearch services running
# on localhost over the CQL API and the Elasticsearch native
# "Transport" API on their respective default ports.  The Cassandra
# and Elasticsearch services must already be running before starting
# JanusGraph with this file.

# The primary persistence provider used by JanusGraph.  This is required.  It
# should be set one of JanusGraph's built-in shorthand names for its standard
# storage backends or to the full package and classname of a custom/third-party
# StoreManager implementation.
#
# Default:    (no default value)
# Data Type:  String
# Mutability: LOCAL
storage.backend =cql

# The hostname or comma-separated list of hostnames of storage backend
# servers.  This is only applicable to some storage backends, such as
# cassandra and hbase.
#
# Default:    127.0.0.1
# Data Type:  class java.lang.String[]
# Mutability: LOCAL
storage.hostname =cassandra.us-east-2.amazonaws.com

# The name of JanusGraph's keyspace.  It will be created if it does not
# exist.
#
# Default:    janusgraph
# Data Type:  String
# Mutability: LOCAL
storage.cql.keyspace =graph

# Whether to enable JanusGraph's database-level cache, which is shared across
# all transactions. Enabling this option speeds up traversals by holding
# hot graph elements in memory, but also increases the likelihood of
# reading stale data.  Disabling it forces each transaction to
# independently fetch graph elements from storage before reading/writing
# them.
#
# Default:    false
# Data Type:  Boolean
# Mutability: MASKABLE
cache.db-cache = true

# How long, in milliseconds, database-level cache will keep entries after
# flushing them.  This option is only useful on distributed storage
# backends that are capable of acknowledging writes without necessarily
# making them immediately visible.
#
# Default:    50
# Data Type:  Integer
# Mutability: GLOBAL_OFFLINE
#
# Settings with mutability GLOBAL_OFFLINE are centrally managed in JanusGraph's
# storage backend.  After starting the database for the first time, this
# file's copy of this setting is ignored.  Use JanusGraph's Management System
# to read or modify this value after bootstrapping.
cache.db-cache-clean-wait = 20

# Default expiration time, in milliseconds, for entries in the
# database-level cache. Entries are evicted when they reach this age even
# if the cache has room to spare. Set to 0 to disable expiration (cache
# entries live forever or until memory pressure triggers eviction when set
# to 0).
#
# Default:    10000
# Data Type:  Long
# Mutability: GLOBAL_OFFLINE
#
# Settings with mutability GLOBAL_OFFLINE are centrally managed in JanusGraph's
# storage backend.  After starting the database for the first time, this
# file's copy of this setting is ignored.  Use JanusGraph's Management System
# to read or modify this value after bootstrapping.
cache.db-cache-time = 180000

# Size of JanusGraph's database level cache.  Values between 0 and 1 are
# interpreted as a percentage of VM heap, while larger values are
# interpreted as an absolute size in bytes.
#
# Default:    0.3
# Data Type:  Double
# Mutability: MASKABLE
cache.db-cache-size = 0.25

# The indexing backend used to extend and optimize JanusGraph's query
# functionality. This setting is optional.  JanusGraph can use multiple
# heterogeneous index backends.  Hence, this option can appear more than
# once, so long as the user-defined name between "index" and "backend" is
# unique among appearances.Similar to the storage backend, this should be
# set to one of JanusGraph's built-in shorthand names for its standard index
# backends (shorthands: lucene, elasticsearch, es, solr) or to the full
# package and classname of a custom/third-party IndexProvider
# implementation.
#
# Default:    elasticsearch
# Data Type:  String
# Mutability: GLOBAL_OFFLINE
#
# Settings with mutability GLOBAL_OFFLINE are centrally managed in JanusGraph's
# storage backend.  After starting the database for the first time, this
# file's copy of this setting is ignored.  Use JanusGraph's Management System
# to read or modify this value after bootstrapping.
index.search.backend =elasticsearch

# The hostname or comma-separated list of hostnames of index backend
# servers.  This is only applicable to some index backends, such as
# elasticsearch and solr.
#
# Default:    127.0.0.1
# Data Type:  class java.lang.String[]
# Mutability: MASKABLE
index.search.hostname = ...

# Or start ES inside the JanusGraph JVM
#index.search.backend=elasticsearch
#index.search.directory=db/es
#index.search.elasticsearch.client-only=false
#index.search.elasticsearch.local-mode=true
vertex.index.name=vertex_index
storage.port=9142
storage.cql.write-consistency-level=LOCAL_QUORUM
storage.cql.ssl.truststore.password=XXX
storage.cql.ssl.truststore.location=/opt/janusgraph/truststore/cassandra_truststore.jks
storage.cql.ssl.enabled=true
storage.cql.replication-factor=3
storage.cql.read-consistency-level=LOCAL_QUORUM
storage.cql.partitioner-name=DefaultPartitioner
storage.cql.only-use-local-consistency-for-system-operations=true
storage.cql.metadata-token-map-enabled=false
storage.cql.metadata-schema-enabled=false
storage.cql.local-datacenter=us-east-2
storage.cql.internal.file-configuration=/opt/janusgraph/conf/cql.conf
storage.buffer-size=2048
storage.batch-loading=true
query.batch=true
metrics.jmx.enabled=true
metrics.jmx.domain=xxx
metrics.jmx.agentid=janusgraph
log.tx.key-consistent=true
log.janusgraph.key-consistent=true
indexing.backend.name=search
index.search.port=443
index.search.elasticsearch.use-mapping-for-es7=true
index.search.elasticsearch.ssl.enabled=true
index.search.elasticsearch.setup-max-open-scroll-contexts=false
index.search.elasticsearch.http.auth.type=custom
index.search.elasticsearch.http.auth.custom.authenticator-class=com.newforma.janusgraph.es.awsauth.AWSV4AuthHttpClientConfigCallback
index.search.elasticsearch.create.ext.number_of_shards=6
index.search.elasticsearch.create.allow-mapping-update=true
index.search.elasticsearch.client-only=true
graph.unique-instance-id=ac1d00021-ca3114a824661
graph.replace-instance-if-exists=true
graph.assign-timestamp=false
edge.index.name=edge_index
channelizer=org.apache.tinkerpop.gremlin.server.channel.WsAndHttpChannelizer
```

**Reply by hexa00** [reply]
Damm just realized double checking that that I have storage.batch-loading=true

**Reply by li-boxuan** [reply]
Hey unfortunately indeed Amazon Keyspace diverges from Cassandra...

https://stackoverflow.com/questions/69940435/amazon-keyspaces-allow-filtering-without-all-keys-in-node-js

They documented it here: https://docs.aws.amazon.com/keyspaces/latest/devguide/working-with-queries.html#paginating-results but I don't think that's enough... ~~IMO it's a shame because they should come up with their own driver library if their server implementation diverges from the standard one.~~

~~We either need to wait for a fix from DataStax java driver (I didn't check if they have already implemented this, but I doubt because DataStax Astra and Amazon Keyspace are competitors), or you have to live with it.~~

After taking a closer look, I have 50% confidence that we can fix it in JanusGraph. I'll create an issue but I don't guarantee when this can be released since reproducing it is difficult. If you'd like to fix it, I would be happy to help.

At the moment, how will this impact you? This will only impact you when you do a full-scan. You should set `query.force-index` to true so that full-scan is not allowed. Of course, this also means you won't be able to do `g.V().count()`. You would have to use an index.

**Reply by li-boxuan** [reply]
I created a bug report on https://github.com/JanusGraph/janusgraph/issues/3390 and also pointed out the direction to a fix. If you have some knowledge of Java and even better, Cassandra java driver, I would recommend you try fixing it and building a JanusGraph release locally (`mvn clean install -Pjanusgraph-release -DskipTests=true --batch-mode --also-make -Dgpg.skip=true`) and testing it out.

**Reply by hexa00** [reply]
Oh very interesting, nice find!!

Yeah I do plan to use indexes only but I was just not understanding why it returned so few results and wanted to make sure things were not broken in ways I did not understand.

I can surely compile and run it... for implementing a fix I'll have to wrap my head around, I did so java and Cassandra code but it's been years..  

I can't imagine the java driver not supporting this seems even cqlsh is aware that there is more results coming even if it receives an empty page.

Could this be affecting other parts of janusgraph or these ALLOW FILTERING queries are only relevant for fullscans ?

**Reply by li-boxuan** [reply]
> Could this be affecting other parts of janusgraph or these ALLOW FILTERING queries are only relevant for fullscans?

Good question. No. JanusGraph only uses ALLOW FILTERING for full scans.

**Reply by hexa00** [reply]
:relieved:  thanks for your help!!!

**Reply by li-boxuan** [reply]
My pleasure and thanks for your patience. Glad we (partially) figured this out together.

**Comment by vtslab** [comment]
I am lost here, but probably others will read this. Can you elaborate on "with our use of keyspaces"?

**Reply by hexa00** [reply]
Oh sorry I can see how this can be confusing I mean we use AWS Keyspaces as our Cassandra backend

**Reply by li-boxuan** [reply]
Did you set a TTL by any chance?

**Reply by hexa00** [reply]
No I enabled however TTL like so:
```
aws keyspaces update-table --keyspace-name graph --table-name edgestore --ttl status=ENABLED;
aws keyspaces update-table --keyspace-name graph --table-name graphindex --ttl status=ENABLED;
aws keyspaces update-table --keyspace-name graph --table-name edgestore_lock_ --ttl status=ENABLED;
aws keyspaces update-table --keyspace-name graph --table-name graphindex_lock_ --ttl status=ENABLED;
aws keyspaces update-table --keyspace-name graph --table-name janusgraph_ids --ttl status=ENABLED;
aws keyspaces update-table --keyspace-name graph --table-name system_properties --ttl status=ENABLED;
aws keyspaces update-table --keyspace-name graph --table-name system_properties_lock_ --ttl status=ENABLED;
aws keyspaces update-table --keyspace-name graph --table-name systemlog --ttl status=ENABLED;
aws keyspaces update-table --keyspace-name graph --table-name txlog --ttl status=ENABLED;
```
And I can see in console:
```
Default Time to Live (TTL) period Info
Never expire (TTL = 0)
```
The fact that I can see all the data doing a query / valueMap suggests the data is there too...?
